<a href="https://colab.research.google.com/github/Davis-Tony/DS-Unit-2-Linear-Models/blob/master/TonyDavis_LS_DS_211_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 2, Sprint 1, Module 1*

---

# Regression 1

## Assignment

You'll use another **New York City** real estate dataset. 

But now you'll **predict how much it costs to rent an apartment**, instead of how much it costs to buy a condo.

The data comes from renthop.com, an apartment listing website.

- [ ] Look at the data. Choose a feature, and plot its relationship with the target.
- [ ] Use scikit-learn for linear regression with one feature. You can follow the [5-step process from Jake VanderPlas](https://jakevdp.github.io/PythonDataScienceHandbook/05.02-introducing-scikit-learn.html#Basics-of-the-API).
- [ ] Define a function to make new predictions and explain the model coefficient.
- [ ] Organize and comment your code.

> [Do Not Copy-Paste.](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit) You must type each of these exercises in, manually. If you copy and paste, you might as well not even do them. The point of these exercises is to train your hands, your brain, and your mind in how to read, write, and see code. If you copy-paste, you are cheating yourself out of the effectiveness of the lessons.

If your **Plotly** visualizations aren't working:
- You must have JavaScript enabled in your browser
- You probably want to use Chrome or Firefox
- You may need to turn off ad blockers
- [If you're using Jupyter Lab locally, you need to install some "extensions"](https://plot.ly/python/getting-started/#jupyterlab-support-python-35)

## Stretch Goals
- [ ] Do linear regression with two or more features.
- [ ] Read [The Discovery of Statistical Regression](https://priceonomics.com/the-discovery-of-statistical-regression/)
- [ ] Read [_An Introduction to Statistical Learning_](http://faculty.marshall.usc.edu/gareth-james/ISL/ISLR%20Seventh%20Printing.pdf), Chapter 2.1: What Is Statistical Learning?

In [1]:
import sys

# If you're on Colab:
if 'google.colab' in sys.modules:
    DATA_PATH = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Applied-Modeling/master/data/'

# If you're working locally:
else:
    DATA_PATH = '../data/'
    
# Ignore this Numpy warning when using Plotly Express:
# FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
import warnings
warnings.filterwarnings(action='ignore', category=FutureWarning, module='numpy')

In [2]:
# Read New York City apartment rental listing data
import pandas as pd
df = pd.read_csv(DATA_PATH+'apartments/renthop-nyc.csv')
assert df.shape == (49352, 34)

In [3]:
# Remove outliers: 
# the most extreme 1% prices,
# the most extreme .1% latitudes, &
# the most extreme .1% longitudes
df = df[(df['price'] >= 1375) & (df['price'] <= 15500) & 
        (df['latitude'] >=40.57) & (df['latitude'] < 40.99) &
        (df['longitude'] >= -74.1) & (df['longitude'] <= -73.38)]

In [4]:
#Looking at dataframe
df.head()

bathrooms  bedrooms  ... wheelchair_access common_outdoor_space
0        1.5         3  ...                 0                    0
1        1.0         2  ...                 0                    0
2        1.0         1  ...                 0                    0
3        1.0         1  ...                 0                    0
4        1.0         4  ...                 0                    0

[5 rows x 34 columns]

In [5]:
#Importing scikit-learn to begin setting up linear regression
from sklearn.linear_model import LinearRegression

In [6]:
#Instantiate this class
model = LinearRegression()

In [9]:
#Arranging feature and target
feature = ['bedrooms']
target = ['price']

x_train = df[feature]
y_train = df[target]

x_train.shape, y_train.shape

((48818, 1), (48818, 1))

In [10]:
#Checking x_train
x_train

bedrooms
0             3
1             2
2             1
3             1
4             4
...         ...
49347         2
49348         1
49349         1
49350         0
49351         2

[48818 rows x 1 columns]

In [11]:
#Checking y_train
y_train

price
0       3000
1       5465
2       2850
3       3275
4       3350
...      ...
49347   3200
49348   3950
49349   2595
49350   3350
49351   2200

[48818 rows x 1 columns]

In [12]:
#Fitting the model
model.fit(x_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [14]:
#Applying model to data
bedrooms = 4
x_test = [[bedrooms]]

y_pred = model.predict(x_test)
y_pred

array([[5680.99081929]])

In [15]:
#Defining function to make predictions
def predict(bedrooms):
  y_pred = model.predict([[bedrooms]])
  estimate = y_pred[0]
  coefficient = model.coef_[0]
  print(int(estimate), "is the estimated price for an apartment with", int(bedrooms), "bedrooms.")
  return estimate

print(predict(5))

6534 is the estimated price for an apartment with 5 bedrooms.
[6534.24498682]


In [16]:
#Explaining the coefficient
model.coef_

array([[853.25416753]])

This shows that each bedroom seems to raise the price of an apartment by roughly $853

In [17]:
#Setting up linear regression with two features
#Recalling df head
df.head()

bathrooms  bedrooms  ... wheelchair_access common_outdoor_space
0        1.5         3  ...                 0                    0
1        1.0         2  ...                 0                    0
2        1.0         1  ...                 0                    0
3        1.0         1  ...                 0                    0
4        1.0         4  ...                 0                    0

[5 rows x 34 columns]

In [18]:
#Arranging new features
feature = ['bedrooms', 'bathrooms']

#no need to reset target because it remains the same, the price
x_train = df[feature]

#no need to set y_train because that has been set from the single feature regression



In [19]:
#Fitting model
model.fit(x_train, y_train)
print(model.coef_)

[[ 385.08418275 2099.14867224]]


In [20]:
#Testing with new data
bedrooms = 3
bathrooms = 2
y_pred = model.predict([[bedrooms, bathrooms]])
y_pred

array([[5818.43891708]])

In [22]:
#Display the coefficients
#importing Plotly Express
import plotly.express as px
fig = px.bar(x=feature, y=model.coef_[0])
fig.show()

Number of bathrooms seems to have more of an effect on the price than do that of bedrooms.